enterprise/server/sociartifactstore/sociartifactstore.go

```diff
@@ -285,112 +327,196 @@ func deserializeDigest(s string) (*repb.Digest, error) {
 
 }
 
-func pullAndIndexImage(ctx context.Context, imageName, imageRef string) (*repb.Digest, []*repb.Digest, error) {
+func (s *SociArtifactStore) pullAndIndexImage(ctx context.Context, imageName string, credentials *rgpb.Credentials) (*repb.Digest, []*repb.Digest, error) {
 	log.Infof("soci artifacts not found, generating them for image %s", imageName)
-	if err := pullImage(ctx, imageName); err != nil {
+	image, err := pullImage(ctx, imageName, credentials)
+	if err != nil {
 		return nil, nil, err
 	}
-	if err := runSoci(ctx, imageName); err != nil {
+	return s.indexImage(ctx, image)
+}
+
+func pullImage(ctx context.Context, image string, credentials *rgpb.Credentials) (v1.Image, error) {
+	imageRef, err := ctrname.ParseReference(image)
+	if err != nil {
+		return nil, status.InvalidArgumentErrorf("invalid image %q", image)
+	}
+
+	remoteOpts := []remote.Option{remote.WithContext(ctx)}
+	if credentials.GetUsername() != "" || credentials.GetPassword() != "" {
+		authenticator := &authn.Basic{
+			Username: credentials.GetUsername(),
+			Password: credentials.GetPassword(),
+		}
+		remoteOpts = append(remoteOpts, remote.WithAuth(authenticator))
+	}
+
+	remoteDesc, err := remote.Get(imageRef, remoteOpts...)
+	if err != nil {
+		if t, ok := err.(*transport.Error); ok && t.StatusCode == http.StatusUnauthorized {
+			return nil, status.PermissionDeniedErrorf("could not retrieve image manifest: %s", err)
+		}
+		return nil, status.UnavailableErrorf("could not retrieve manifest from remote: %s", err)
+	}
+
+	switch remoteDesc.MediaType {
+	case types.OCIImageIndex, types.DockerManifestList:
+		return nil, status.InvalidArgumentErrorf("image index not expected in conversion request")
```

Comment:
`getTargetImageRef` was responsible for checking if the target ref was an image index and then finding a single image to use. you're now passing the image directly to this function which could end up being an image index I believe?

Reply:
Ahh good call, returned to the caller. LMK if you can think of a cleaner way to manage these different components of the image.